docs/core/concepts/results.md

```diff
@@ -28,44 +20,116 @@ An instantiated Result object has the following attributes:
+
+```python
+>>> state.result[first_result]._result.value                 
+3
+>>> state.result[first_result]._result.safe_value            
+<SafeResult: '3'>
+```
+
+In Core, the return value is associated in-memory with the `Result` object attached to flow `State`s. For Cloud customers, the return value of a result handler's `write` method is also persisted in Prefect Cloud's database.
 
 ::: warning Handle your data carefully
-When running on Prefect Cloud, the output of a Result Handler's `write` method is what is stored in the Cloud database.
+Using result handlers means that data will be persisted beyond the Prefect's Python process to a storage location, so it is worth taking some extra time to consider what data is safe to persist where.
+
+When running on Prefect Cloud, the return value of a result handler's `write` method is what is stored in the Cloud database. Though you may serialize your data during your result handler's execution, *only* `return` data from Tasks that you feel is safe to persist in Prefect Cloud.
 :::
 
 ### How to specify a `ResultHandler`
 
-There is a hierarchy to determining what `ResultHandler` to use for a given piece of data:
+`ResultHandler`'s `read` and `write` methods will be called by Prefect only if you enable checkpointing. Once this configuration is set, there is a hierarchy to determining what `ResultHandler` to use for a given piece of data:
 
-1. First, users can set a global default in their Prefect user config; if you never mention or think about Result Handlers again, this is the handler that will always be used.
-1. Next, you can specify a Flow-level result handler at Flow-initialization using the `result_handler` keyword argument. Once again, if you never specify another result handler, this is the one that will be used for all your tasks in this particular Flow.
+1. You can specify a Flow-level result handler at Flow-initialization using the `result_handler` keyword argument. If you never specify another result handler, this is the one that will be used for all your tasks in this particular Flow.
 1. Lastly, you can set a Task-level result handler. This is achieved using the `result_handler` keyword argument at Task initialization (or in the `@task` decorator). If you provide a result handler here, it will _always_ be used if the _output_ of this Task needs to be cached for any reason whatsoever.
 
 ::: tip The Hierarchy
-Task-level result handlers will _always_ be used over Flow-level handlers; Flow-level result handlers will _always_ be used over the handler set in your config.
+Task-level result handlers will _always_ be used over flow-level handlers. Neither will be used if a task's `checkpointing` kwarg is set to `False`, or the global `prefect.config.tasks.checkpointing` value is set to `False`.
+:::
+
+::: warning result handlers are always attached to task outputs
```

Comment:
@cicdw this tip box is where I'm stuck with the word `cache` that I'm not sure what it means. If B is doing input caching, my understanding now is that should be handled in-process by saving a python object in `CachedStates`, without touching A's result handler. 

Also not 100% sure the capitalization situation for the title is to standard, may have been over-grepped unless this is correct

Reply:
Yea I think you should at least capitalize the `r`; you are correct when running entirely in Core.  In Cloud, A's result handler will be called so that the `cached_inputs` attribute of B's `CachedState` is also safe

Reply:
Updated in https://github.com/PrefectHQ/prefect/pull/2076/commits/6ecd02bc9b0a23947b40ad8249258b781a4a1260 also with a reference to the cloud docs that state result handlers are called in this situation